File: Roblox/Roblox.Libraries/Assets/Imager.cs
using NetVips;

namespace Roblox.Libraries;

public enum ImagerFormat
{
    Undefined = 0,
    PNG = 1,
    JPEG,
    GIF,
    BMP,
}

public class UnsupportedImageFormatException : Exception
{

}

/// <summary>
/// Generic parse error, e.g. image format is not supported, image is corrupted, image is invalid
/// </summary>
public class InvalidImageException : Exception
{

}

public class Imager
{
    private Stream content { get; set; }

    public int height { get; private set; } = 0;
    public int width { get; private set; } = 0;
    public ImagerFormat imageFormat { get; private set; } = ImagerFormat.Undefined;

    private Image? image { get; set; }

    private Imager(Stream content)
    {
        this.content = content;
    }

    private Task InitializeAsync()
    {
        Image imageData;
        ImagerFormat decodedFormat;
        try
        {
            var loader = Image.FindLoadStream(content);
            if (content.CanSeek)
                content.Position = 0;

            imageData = Image.NewFromStream(content, "", access: Enums.Access.Sequential, failOn: Enums.FailOn.Error);
            decodedFormat = GetImageFormat(loader);
        }
        catch (VipsException)
        {
            throw new InvalidImageException();
        }

        this.image = imageData;
        height = imageData.Height;
        width = imageData.Width;
        imageFormat = decodedFormat;
        return Task.CompletedTask;
    }

    public static async Task<Imager> ReadAsync(Stream content)
    {
        var img = new Imager(content);
        await img.InitializeAsync();
        return img;
    }

    private static ImagerFormat GetImageFormat(string? loader)
    {
        if (loader == null)
            throw new InvalidImageException();

        if (loader.Contains("png", StringComparison.OrdinalIgnoreCase))
            return ImagerFormat.PNG;
        if (loader.Contains("jpeg", StringComparison.OrdinalIgnoreCase))
            return ImagerFormat.JPEG;
        if (loader.Contains("gif", StringComparison.OrdinalIgnoreCase))
            return ImagerFormat.GIF;
        if (loader.Contains("bmp", StringComparison.OrdinalIgnoreCase))
            return ImagerFormat.BMP;

        throw new UnsupportedImageFormatException();
    }
}

File: game-renderer/src/Utilities/Libraries/Imager.ts
import sharp, {Sharp} from "sharp";
import {Console} from "./CS.js";

export class Imager {
    private content: Buffer;
    public height = 0;
    public width = 0;
    public aspectRatio = 0;
    public imageFormat: ImagerFormat = ImagerFormat.Undefined;

    private sharp: sharp.Sharp | null = null;
    private metadata: sharp.Metadata | null = null;

    private constructor(content: Buffer) {
        this.content = content;
    }

    public GetImage(): Sharp | null {
        return this.sharp;
    }

    public GetMetadata(): sharp.Metadata | null {
        return this.metadata;
    }

    private async InitializeAsync(): Promise<void> {
        try {
            this.sharp = sharp(this.content);
            this.metadata = await this.sharp.metadata();
        } catch (e) {
            Console.Error("Invalid image provided");
            return Promise.resolve();
        }

        if (!this.metadata || !this.metadata.format) {
            Console.Error("Invalid image provided");
            return  Promise.resolve();
        }

        this.height = this.metadata.height || 0;
        this.width = this.metadata.width || 0;
        this.aspectRatio = this.width !== 0 && this.height !== 0 ? this.width / this.height : 0;

        switch (this.metadata.format.toLowerCase()) {
            case "png":
                this.imageFormat = ImagerFormat.PNG;
                break;
            case "jpeg":
            case "jpg":
                this.imageFormat = ImagerFormat.JPEG;
                break;
            case "gif":
                this.imageFormat = ImagerFormat.GIF;
                break;
            case "bmp":
                this.imageFormat = ImagerFormat.BMP;
                break;
            default:
                Console.Error("Invalid image provided");
                return  Promise.resolve();
        }
    }

    public static async ReadAsync(content: Buffer): Promise<Imager> {
        const img = new Imager(content);
        await img.InitializeAsync();
        return img;
    }
}

export enum ImagerFormat {
    Undefined = 0,
    PNG,
    JPEG,
    GIF,
    BMP,
}
